rfc/rfc-0009-controller-unit-testing.md

```diff
@@ -0,0 +1,83 @@
+# RFC-0009: Controller Unit Testing
+
+**Authors:** Glyn Normington
+
+**Status:**
+
+**Pull Request URL:** https://github.com/projectriff/riff/pull/1372
+
+**Superseded by:** N/A
+
+**Supersedes:** N/A
+
+**Related:** N/A
+
+
+## Problem
+The Kubernetes controllers in riff need unit testing in a consistent way. Other projects
+have adopted and promoted various ways of testing controllers. For ease of maintenance,
+riff needs to adopt a single, consistent approach.
+
+For the purpose of this document, unit testing refers to testing the
+`Reconcile` method of a controller.
+
+Apart from checking that various objects are created etc., it should be straigthtforward
+to test error paths by injecting errors in object lookup and creation.
+
+### Anti-Goals
+
+None.
+
+## Solution
+
+The approach taken is to use a fake client and other interfaces and introduce a test framework
+which allows tests to be written in a declarative style. See "Rejected Alternatives" below for 
+some other approaches which were tried out.
+
+The test framework supports table-driven tests. Each row of the table tests a single call to `Reconcile` by
+setting up some initial Kubernetes objects, modifying the client behaviour using various hooks
+or reactors, and then asserting that various Kubernetes objects are created, updated, etc.
+
+When a helper method is used to construct a Kubernetes object, it should construct a fixed value and any variations
+should be explicitly coded in the table row. This avoids details of a particular test being "buried" in helper methods.
+
+Since kubebuilder was used to generate the controllers, some packages contain more than one controller and so
+helper methods need to be named or organised carefully to avoid clashes. 
+
+## Rejected Alternatives
+
+### Testing with a Local Control Plane
+riff's controllers were generated by kubebuilder and depend on the controller
+runtime. Kubebuilder recommends integration testing using a local control plane.
+
+This was rejected because the control plane
+(specifically `etcd`) sometimes fails to start, which causes intermittent test failures.
+Such failures are a time sink when developers, particularly newcomers, end up diagnosing
+false positives, especially in CI.
+
+A local control plane also complicates tests because
+deletion of Kubernetes objects is not synchronous and so this needs to be worked around by
+generating distinct object names or by other means.
+
+Also, a local control plane is not
+deterministic when it generates object names, which complicates the setting up of test
+expectations.
+
+See here for a quick
+[spike](https://github.com/projectriff/system/pull/232) using a local control plane.
+
+### Depending on Knative testing packages
+Another approach which was quickly [spiked](https://github.com/projectriff/system/pull/233) and
+rejected was to depend on Knative testing packages. This was abandoned because of dependency
```

Comment:
Knative also uses typed clientsets, which is not the same client interface we're using from controller runtime.

Reply:
Good point. Changed.